### PR TITLE
Batch2

### PR DESCRIPTION
**Pull request by TristanBilot**
This is a duplicated branch of @Kheims' work for batching as the first branch was broken.
I integrated batch for arbitrary array attributes. `__cat_dim__` and `__incr__` have been implemented and tested. Slicing and indexing with negative numbers is also done. 

I will add docs, edge case errors and more tests tomorrow.

### CODE REVIEW

File: mlx_graphs/data/batch.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+from typing import List, Union
+import mlx.core as mx
+
+from mlx_graphs.data.data import GraphData
+from mlx_graphs.data.collate import collate
+
+
+class GraphDataBatch(GraphData):
+    def __init__(self, graphs, **kwargs) -> None:
```
Comment (francescofarina) [inline mlx_graphs/data/batch.py:9]:
```suggestion
    def __init__(self, graphs: list[GraphData], **kwargs) -> None:
```

File: mlx_graphs/data/batch.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
+from typing import List, Union
```
Comment (francescofarina) [inline mlx_graphs/data/batch.py:1]: Importing `List` is not needed anymore - we can directly use `list` (see PEP484 or [here](https://docs.python.org/3/library/typing.html)).

File: mlx_graphs/data/collate.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+from typing import List
```
Comment (francescofarina) [inline mlx_graphs/data/collate.py:1]: As above, no need for `List`

File: mlx_graphs/data/batch.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+            # If __inc__ is True, we get back the original values by subtracting
+            # the cumulative sum at the batch index
+            if self.to_dict()[f"_inc_{attr}"]:
+                original_value -= self._cumsum[idx]
+
+            # NOTE: `self._cumsum` is the cumsum of number of nodes per graph. For custom features
+            # with __inc__, each feature should have its own cumsum
+
+            single_graph_dict[attr] = original_value
```
Comment (francescofarina) [inline mlx_graphs/data/batch.py:73]: this is very neat!

File: mlx_graphs/data/collate.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    apart from `edge_index` which is concatenated along dim 1 and
+    nodes indices are incremented
+
+    Args:
+        graph_list (List[GraphData]): the list of GraphData objects to collate
+
+    Returns:
+        dict: dict containing all the attributes of the unified and disconnected big graph as
+        well as the slices corresponding to the portion of node for each graph from the provided
+        list of graphs.
```
Comment (francescofarina) [inline mlx_graphs/data/collate.py:20]: No need for type specs in docstrings as they're now inferred from the signature

```suggestion
    Args:
        graph_list: the list of GraphData objects to collate

    Returns:
        Dict containing all the attributes of the unified and disconnected big graph as
        well as the slices corresponding to the portion of node for each graph from the provided
        list of graphs.
```

File: mlx_graphs/data/collate.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    batch_attr_dict = {}
+
+    for attr in graph_list[0].to_dict():
```
Comment (francescofarina) [inline mlx_graphs/data/collate.py:24]: This is very much an edge case, but what if some graphs in the list have additional attributes? Should they be concatenated in some way? Ignored? Should we throw an error?

**Review by francescofarina — CHANGES_REQUESTED** [review]
Just some minor initial comments! I'll re-review after docs and full tests are implemented

File: mlx_graphs/data/collate.py (hunk 6 above)
Comment (TristanBilot) [reply mlx_graphs/data/collate.py:24]: Ahha that's exactly what I planed to work on today. I think we should either ignore other attributes, or raise an error. 
Just ignoring might by tricky for the user if he's looking for those values in the final batch, and raising might be a bit too strong, but I don't see why some GrahData would have additional attributes.

Comment (francescofarina) [reply mlx_graphs/data/collate.py:24]: Yeah, I agree that's something that shouldn't happen, at least for homogeneous graphs - it could if they're heterogeneous though, but we can leave that for later!

I think in this case we should raise an error as having extra attributes is something we don't expect, or we can raise a warning saying that extra attributes will be ignored, wdyt?

Comment (TristanBilot) [reply mlx_graphs/data/collate.py:24]: Okay let's go for the raising option!

Comment (francescofarina) [reply mlx_graphs/data/collate.py:24]: The only thing I'd be slightly worried about is the overhead we'll incur in checking all the graphs in the batch - but it should be a fairly cheap operation.

File: docs/source/api/data/batch.rst [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+Batch Objects
+=============
+
+.. currentmodule:: mlx_graphs.data.batch
+
+
+.. autosummary::
+	:toctree: _autosummary
+
+	GraphDataBatch
```
Comment (TristanBilot) [inline docs/source/api/data/batch.rst:12]: @francescofarina do you know how to consider `__incr__` and `__cat_dim__` methods in the docs?

Comment (francescofarina) [reply docs/source/api/data/batch.rst:12]: We'll likely have to write a custom template (right now I wrote only this docs/source/_templates/custom-class-template.rst for the nn module to include the `__call__`). Why do you want to include them in the doc?

Comment (TristanBilot) [reply docs/source/api/data/batch.rst:12]: It's important to have these in the docs as the user can override them for custom batching.

Comment (francescofarina) [reply docs/source/api/data/batch.rst:12]: Ok, then a custom template is the way to go.

File: mlx_graphs/data/batch.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
+from typing import Union
+import mlx.core as mx
+
+from mlx_graphs.data.data import GraphData
+from mlx_graphs.data.collate import collate
+
+
+class GraphDataBatch(GraphData):
+    """Concatenates multiple `GraphData` into a single unified `GraphBatch`
```
Comment (francescofarina) [inline mlx_graphs/data/batch.py:9]: `GraphBatch` -> `GraphDataBatch`

File: mlx_graphs/data/batch.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+                node_features=array([[1], [1], [1]], dtype=float32)),
+            GraphData(
+                edge_index=array([[3, 3, 3], [4, 4, 4]], dtype=int32),
+                node_features=array([[2], [2], [2]], dtype=float32))
+        ]
+    """
+
+    def __init__(self, graphs: list[GraphData], **kwargs) -> None:
+        if not isinstance(graphs, (list, tuple)):
+            graphs = list(graphs)
```
Comment (francescofarina) [inline mlx_graphs/data/batch.py:66]: I think we can safely remove this as it's clearly documented that the argument needs to be a list of graph. We can eventually add the decorator `@validate_list_of_graph_data`

File: mlx_graphs/data/batch.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    list of :class:`~mlx_graphs.data.GraphData`
+    """
+    return GraphDataBatch(graphs)
+
+
+def unbatch(batch: GraphDataBatch) -> list[GraphData]:
+    """Reconstructs a list of :class:`~mlx_graphs.data.GraphData`
+    objects from the :class:`~mlx_graphs.data.GraphDataBatch` object.
+    """
+    return [batch[idx] for idx in range(batch.num_graphs)]
```
Comment (francescofarina) [inline mlx_graphs/data/batch.py:167]: update docs

File: docs/source/api/data/batch.rst (hunk 7 above)
Comment (TristanBilot) [reply docs/source/api/data/batch.rst:12]: ok we'll handle that in another PR

File: mlx_graphs/data/collate.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    apart from `edge_index` which is concatenated along dim 1.
+    Each graph remains independent in the final graph by incrementing
+    the indices in `edge_index` based on the cumsum of previous number
+    of nodes per graph.
+
+    Args:
+        graph_list: List of `GraphData` objects to collate
+
+    Returns:
+        dict: dict containing all the attributes of the unified and disconnected big graph as
```
Comment (francescofarina) [inline mlx_graphs/data/collate.py:22]:
```suggestion
        Dict containing all the attributes of the unified and disconnected big graph as
```

File: mlx_graphs/data/data.py [hunk 12]
```diff
@@ -47,19 +47,61 @@ def __init__(
...
+
+        By default, all attribute names containing "index" will be concatenated on axis 1,
+        e.g. `edge_index`. Other attributes are concatenated on axis 0, e.g. node features.
+
+        Args:
+            key: Name of the attribute on which change the default concatenation dimension
+                while using batching.
+
+        Returns:
+            int: The dimension where concatenation will happen when batching.
```
Comment (francescofarina) [inline mlx_graphs/data/data.py:82]:
```suggestion
            The dimension where concatenation will happen when batching.
```

File: mlx_graphs/data/data.py [hunk 13]
```diff
@@ -47,19 +47,61 @@ def __init__(
...
+        By default, all attribute names containing "index" will be incremented to avoid
+        duplicate nodes in the index, e.g. `edge_index`.
+        Other attributes are cnot incremented and keep their original values, e.g. node features.
+
+        Args:
+            key: Name of the attribute on which change the default incrementation behavior
+                while using batching.
+
+        Returns:
+            bool: Whether to use incrementation of values for a given attribute.
```
Comment (francescofarina) [inline mlx_graphs/data/data.py:103]:
```suggestion
            Whether to use incrementation of values for a given attribute.
```

File: tests/nn/test_message_passing.py [hunk 14]
```diff
@@ -59,9 +59,9 @@ def test_sum_aggregation():
     y3 = mx.array([[2, 4], [5, 6], [0, 0]])
     y_hat3 = mpnn(x, edge_index, edge_weight=None)
 
-    edge_index = mx.array([[0, 100, 0], [0, 1, 100]])
-    y4 = mx.array([[1, 2], [0, 0], [0, 0]])
-    y_hat4 = mpnn(x, edge_index, edge_weight=None)
+    # edge_index = mx.array([[0, 100, 0], [0, 1, 100]])
+    # y4 = mx.array([[1, 2], [0, 0], [0, 0]])
+    # y_hat4 = mpnn(x, edge_index, edge_weight=None)
```
Comment (francescofarina) [inline tests/nn/test_message_passing.py:64]: What error do you get here with `pytest`? Looks like it works ok for me

**Review by francescofarina — CHANGES_REQUESTED** [review]

File: tests/nn/test_message_passing.py (hunk 14 above)
Comment (francescofarina) [reply tests/nn/test_message_passing.py:64]: No, now failing for me as well

File: mlx_graphs/data/batch.py (hunk 9 above)
Comment (TristanBilot) [reply mlx_graphs/data/batch.py:66]: Ok, `collate()` already handles `@validate_list_of_graph_data` so no need for an additional call.

File: mlx_graphs/data/batch.py (hunk 10 above)
Comment (TristanBilot) [reply mlx_graphs/data/batch.py:164]: Do we need `:class:~mlx_graphs.data.GraphData` here or we still use `GraphData` like everywhere else.

Comment (francescofarina) [reply mlx_graphs/data/batch.py:164]: I don't think we need it, especially if you add `args` and `returns`

File: tests/nn/test_message_passing.py (hunk 14 above)
Comment (TristanBilot) [reply tests/nn/test_message_passing.py:64]: Yes it's a strange one

**Comment by francescofarina** [comment]
@TristanBilot I added the docs template here https://github.com/TristanBilot/mlx-graphs/pull/41  - it will be merged into this branch

**Comment by TristanBilot** [comment]
Ok nice, you can merge it into this one when you're done.

**Comment by TristanBilot** [comment]
I just implemented `batch_indices`, which is similar to `batch` in PyG.

**Comment by TristanBilot** [comment]
I got few errors with the docs:
`mlx_graphs.data.batch.GraphDataBatch.rst: WARNING: document isn't included in any toctree`
`mlx_graphs.data.batch.GraphDataBatch, use :no-index: for one of them`

**Comment by francescofarina** [comment]
> I got few errors with the docs: `mlx_graphs.data.batch.GraphDataBatch.rst: WARNING: document isn't included in any toctree` `mlx_graphs.data.batch.GraphDataBatch, use :no-index: for one of them`

Try deleting the `docs/build` and removing the generated `_autosummary*` subfolders (in source) before compiling the docs, you may have some outdated files. It works well for me

File: mlx_graphs/data/data.py [hunk 15]
```diff
@@ -47,19 +47,62 @@ def __init__(
...
 
         Returns:
-            dict: A dictionary representation of the Data object.
+            A dictionary with all attributes of the `GraphData` object.
         """
         return {k: v for k, v in self.__dict__.items() if v is not None}
+
+    @property
+    def num_nodes(self) -> int:
+        return self.node_features.shape[0]
```
Comment (francescofarina) [inline mlx_graphs/data/data.py:67]: `self.node_features` may be None

```suggestion
    @property
    def num_nodes(self) -> Optional[int]:
        if self.node_features:
            return self.node_features.shape[0]
        return None
```

File: mlx_graphs/data/data.py [hunk 16]
```diff
@@ -47,19 +47,62 @@ def __init__(
...
+
+        Args:
+            key: Name of the attribute on which change the default incrementation behavior
+                while using batching.
+
+        Returns:
+            Incrementing value for the given attribute or None.
+        """
+        if "index" in key:
+            return len(self.node_features)
```
Comment (francescofarina) [inline mlx_graphs/data/data.py:107]: `self.node_features` may be None

**Review by francescofarina — CHANGES_REQUESTED** [review]
Just two extra comments in `data.py` as `self.node_features` may be None

File: mlx_graphs/data/data.py (hunk 16 above)
Comment (francescofarina) [reply mlx_graphs/data/data.py:107]:
```suggestion
            return len(self.node_features) if self.node_features else None
```

File: mlx_graphs/data/data.py (hunk 15 above)
Comment (TristanBilot) [reply mlx_graphs/data/data.py:67]: The None node_features is unexpected behavior for the moment, we didn't handle this case in messsage passing yet. 
I think I will have to do other modifications in the project to handle this case properly. 
Are node features also optional in PyG?

Comment (francescofarina) [reply mlx_graphs/data/data.py:67]: Yes they are - see [here](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/data/data.html#Data). In general all attributes are optional so we'll have to handle that

Comment (francescofarina) [reply mlx_graphs/data/data.py:67]: We can also do that in another PR btw.

**Comment by TristanBilot** [comment]
I think I'm done with this PR, I'll add more handling with None node features later in another PR

**Comment by francescofarina** [comment]
Cool, merging now
